x/autopilot/keeper/liquidstake.go

```diff
@@ -16,103 +15,141 @@ import (
 	}
 
 	if err := msg.ValidateBasic(); err != nil {
 		return err
 	}
 
 	msgServer := stakeibckeeper.NewMsgServerImpl(k.stakeibcKeeper)
-	result, err := msgServer.LiquidStake(
+	msgResponse, err := msgServer.LiquidStake(
 		sdk.WrapSDKContext(ctx),
 		msg,
 	)
 	if err != nil {
 		return errorsmod.Wrapf(err, err.Error())
 	}
 
-	if packetMetadata.IbcReceiver == "" {
+	// If the IBCReceiver is empty, there is no forwarding step
+	if autopilotMetadata.IbcReceiver == "" {
 		return nil
 	}
 
-	hostZone, err := k.stakeibcKeeper.GetHostZoneFromHostDenom(ctx, token.Denom)
+	// Otherwise, if there is forwarding info, submit the IBC transfer
+	return k.IBCTransferStToken(ctx, msgResponse.StToken, transferMetadata, autopilotMetadata)
+}
+
+// Submits an IBC transfer of the stToken to a non-stride zone (either back to the host zone or to a different zone)
+// The sender of the transfer is the hashed receiver of the original autopilot inbound transfer
+func (k Keeper) IBCTransferStToken(
+	ctx sdk.Context,
+	stToken sdk.Coin,
+	transferMetadata transfertypes.FungibleTokenPacketData,
+	autopilotMetadata types.StakeibcPacketMetadata,
+) error {
+	hostZone, err := k.stakeibcKeeper.GetHostZoneFromHostDenom(ctx, transferMetadata.Denom)
 	if err != nil {
 		return err
 	}
 
-	return k.IBCTransferStAsset(ctx, result.StToken, addr.String(), hostZone, packetMetadata)
-}
-
-func (k Keeper) IBCTransferStAsset(ctx sdk.Context, stAsset sdk.Coin, sender string, hostZone *stakeibctypes.HostZone, packetMetadata types.StakeibcPacketMetadata) error {
-	ibcTransferTimeoutNanos := k.stakeibcKeeper.GetParam(ctx, stakeibctypes.KeyIBCTransferTimeoutNanos)
-	timeoutTimestamp := uint64(ctx.BlockTime().UnixNano()) + ibcTransferTimeoutNanos
-	channelId := packetMetadata.TransferChannel
+	// If there's no channelID specified in the packet, default to the channel on the host zone
+	channelId := autopilotMetadata.TransferChannel
 	if channelId == "" {
 		channelId = hostZone.TransferChannelId
 	}
+
+	// Generate a hashed address for the sender to the next hop,
+	// to prevent impersonation at downstream zones
+	// Note: The channel ID here is different than the one used in PFM
+	// (we use the outbound channelID, they use the inbound channelID)
+	// DOUBLE CHECK ME that it shouldn't matter
+	hashedAddress, err := types.GenerateHashedAddress(channelId, transferMetadata.Sender)
+	if err != nil {
+		return err
+	}
+
+	// First we need to bank send to the hashed address
+	originalReceiver, err := sdk.AccAddressFromBech32(transferMetadata.Receiver)
+	if err != nil {
+		return err
+	}
+	hashedSender, err := sdk.AccAddressFromBech32(hashedAddress)
+	if err != nil {
+		return err
+	}
+	if err := k.bankKeeper.SendCoins(ctx, originalReceiver, hashedSender, sdk.NewCoins(stToken)); err != nil {
+		return err
+	}
+
+	// Use the default transfer timeout of 10 minutes
+	timeoutTimestamp := uint64(ctx.BlockTime().UnixNano()) + transfertypes.DefaultRelativePacketTimeoutTimestamp
```

Comment:
how are you thinking about setting the timeout?

my default assumption would have been that longer timeouts (on the order of hours) are better

Reply:
My thinking was two fold:
* (1) Users are used to a 10minute timeout for transfers. If I was waiting longer than that and my tokens were in limbo, it would make me a bit uneasy
* (2) Timeouts on transfers are very rare. I don't think any of our epochly transfers has ever timed out and those have a 30minute timeout

Not opposed to making it longer, but that was my thought process!

Reply:
That makes sense! For what it's worth, I've only seen transfers time out in periods of high activity on Osmosis. Two that come to mind - (1) Terra collapse (2) Celestia launch. In both cases, it took me about ~2 hours to get an IBC transfer processed.

In a bull market, we could experience events of this magnitude monthly.

If we end up retrying forwards on timed out packets, it doesn't matter too much what the timeout is. If we limit the retries, I'd bias to either a longer timeout or allowing for more retries

Reply:
I also experienced the timeouts with the Celestia launch, which is partially what motivated this sentiment. (I found myself a bit anxious while I was waiting haha)

Do you think either approach introduces more complexity on the frontend based on how long we cache a users session?

Reply:
Chatted offline, landed on 3 hour timeouts

Reply:
updated to 3 hours [[here](https://github.com/Stride-Labs/stride/pull/1038/commits/2bf291f924a1587cbb8092237f4602e9d8e3830f)]